tools/rtpsrelay/lib/QosIndex.h

```diff
@@ -0,0 +1,1145 @@
+        } else {
+          PatternType* pattern = node_.find_pattern(name);
+          if (pattern == nullptr) {
+            pattern = new PatternType();
+            node_.insert(name, pattern);
+            for (LiteralType* literal : node_.find_literals(name)) {
+              pattern->insert(literal);
+            }
+          }
+          pattern->insert(reader);
+        }
+      }
+    }
+  }
+
+  void erase(Reader* reader)
+  {
+    const DDS::PartitionQosPolicy& qos = reader->reader_entry._subscriber_qos.partition;
+    if (qos.name.length() == 0) {
+      index_.erase(reader);
+      return;
+    }
+
+    for (int idx = 0, limit = qos.name.length(); idx != limit; ++idx) {
+      Name name(qos.name[idx].in());
+      if (!name.is_valid()) {
+        continue;
+      }
+      if (name.is_literal()) {
+        LiteralType* literal = node_.find_literal(name);
+        literal->erase_literal(reader);
+        if (literal->empty()) {
+          node_.remove(name);
+          delete literal;
+        }
+      } else {
+        PatternType* pattern = node_.find_pattern(name);
+        pattern->erase(reader);
+        if (pattern->empty()) {
+          node_.remove(name);
+          delete pattern;
+        }
+      }
+    }
+  }
+
+  bool empty() const { return node_.empty() && index_.empty(); }
+
+  void get_readers(Writer* writer, ReaderSet& readers) const
+  {
+    const DDS::PartitionQosPolicy& qos = writer->writer_entry.publisher_qos().partition;
+    if (qos.name.length() == 0) {
+      index_.get_readers(writer, readers);
+      return;
+    }
+
+    for (int idx = 0, limit = qos.name.length(); idx != limit; ++idx) {
+      Name name(qos.name[idx].in());
+      if (!name.is_valid()) {
+        continue;
+      }
+      if (name.is_literal()) {
+        LiteralType* literal = node_.find_literal(name);
+        if (literal) {
+          literal->get_readers(writer, readers);
+        }
+      } else {
+        PatternType* pattern = node_.find_pattern(name);
+        if (pattern) {
+          pattern->get_readers(writer, readers);
+        }
+      }
+    }
+  }
+
+  void get_writers(Reader* reader, WriterSet& writers) const
+  {
+    const DDS::PartitionQosPolicy& qos = reader->reader_entry.subscriber_qos().partition;
+    if (qos.name.length() == 0) {
+      index_.get_writers(reader, writers);
+      return;
+    }
+
+    for (int idx = 0, limit = qos.name.length(); idx != limit; ++idx) {
+      Name name(qos.name[idx].in());
+      if (!name.is_valid()) {
+        continue;
+      }
+      if (name.is_literal()) {
+        LiteralType* literal = node_.find_literal(name);
+        if (literal) {
+          literal->get_writers(reader, writers);
+        }
+      } else {
+        PatternType* pattern = node_.find_pattern(name);
+        if (pattern) {
+          pattern->get_writers(reader, writers);
+        }
+      }
+    }
+  }
+
+private:
+  TrieNode<SubIndex> node_;
+  SubIndex index_; // For no partitions.
```

Comment:
"The zero-length
sequence is treated as a special value
equivalent to a sequence containing a single
element consisting of the empty string."